numba/npyufunc/parfor.py

```diff
@@ -0,0 +1,533 @@
+    for label, block in gufunc_ir.blocks.items():
+        for i, inst in enumerate(block.body):
+            if isinstance(inst, ir.Assign) and inst.target.name=="__sentinel__":
+                # We found the sentinel assignment.
+                loc = inst.loc
+                scope = block.scope
+                # split block across __sentinel__
+                # A new block is allocated for the statements prior to the sentinel
+                # but the new block maintains the current block label.
+                prev_block = ir.Block(scope, loc)
+                prev_block.body = block.body[:i]
+                # The current block is used for statements after the sentinel.
+                block.body = block.body[i+1:]
+                # But the current block gets a new label.
+                body_first_label = min(loop_body.keys())
+                # The previous block jumps to the minimum labelled block of the
+                # parfor body.
+                prev_block.append(ir.Jump(body_first_label, loc))
+                # Add all the parfor loop body blocks to the gufunc function's IR.
+                for (l, b) in loop_body.items():
+                    gufunc_ir.blocks[l] = b
+                body_last_label = max(loop_body.keys())
+                gufunc_ir.blocks[new_label] = block
+                gufunc_ir.blocks[label] = prev_block
+                # Add a jump from the last parfor body block to the block containing
+                # statements after the sentinel.
+                gufunc_ir.blocks[body_last_label].append(ir.Jump(new_label, loc))
+                break
+        else:
+            continue
+        break
+
+    gufunc_ir.blocks = numba.parfor._rename_labels(gufunc_ir.blocks)
+    remove_dels(gufunc_ir.blocks)
+
+    if config.DEBUG_ARRAY_OPT:
+        print("gufunc_ir last dump")
+        gufunc_ir.dump()
+
+    kernel_func = compiler.compile_ir(typingctx, targetctx, gufunc_ir, gufunc_param_types, types.none, flags, locals)
+
+    kernel_sig = signature(types.none, *gufunc_param_types)
+    if config.DEBUG_ARRAY_OPT:
+        print("kernel_sig = ", kernel_sig)
+
+    return kernel_func, parfor_args, kernel_sig
+
+
+def call_parallel_gufunc(lowerer, cres, gu_signature, outer_sig, expr_args, loop_ranges, array_size_vars, redvars, reddict, init_block):
+    '''
+    Adds the call to the gufunc function from the main function.
+    '''
+    context = lowerer.context
+    builder = lowerer.builder
+    library = lowerer.library
+
+    from .parallel import ParallelGUFuncBuilder, build_gufunc_wrapper, get_thread_count, _launch_threads, _init
+
+    if config.DEBUG_ARRAY_OPT:
+        print("make_parallel_loop")
+        print("args = ", expr_args)
+        print("outer_sig = ", outer_sig.args, outer_sig.return_type, outer_sig.recvr, outer_sig.pysig)
+
+    # Build the wrapper for GUFunc
+    ufunc = ParallelGUFuncBuilder(cres.entry_point, gu_signature)
```

Comment:
It seems that this ufunc object and the following operation on it is unnecessary.